File: serviceos-ios-core/Sources/ServiceOSIOSCore/Context.swift
public struct ServiceContext: Sendable, Equatable {
    public let contextId: String
    public let contextVersion: String
    public let scopeRef: String?

    public init(contextId: String, contextVersion: String, scopeRef: String? = nil) {
        precondition(!contextId.isEmpty && !contextVersion.isEmpty, "服务上下文缺少稳定标识或版本")
        self.contextId = contextId
        self.contextVersion = contextVersion
        self.scopeRef = scopeRef
    }
}

public actor ContextSelectionStore {
    private var selected: ServiceContext?
    private let onBoundaryChanged: @Sendable () async -> Void

    public init(onBoundaryChanged: @escaping @Sendable () async -> Void) {
        self.onBoundaryChanged = onBoundaryChanged
    }

    public func current() -> ServiceContext? { selected }

    public func select(_ next: ServiceContext) async {
        if let selected, selected.contextId != next.contextId || selected.contextVersion != next.contextVersion {
            await onBoundaryChanged()
        }
        selected = next
    }

    public func clear() async {
        if selected != nil { await onBoundaryChanged() }
        selected = nil
    }
}

File: serviceos-web-core/src/context.ts
export type ServiceContext = Readonly<{
  contextId: string
  contextVersion: string
  scopeRef?: string
  [key: string]: unknown
}>

export type ContextSelectionStore<T extends ServiceContext> = Readonly<{
  current(): T | null
  select(next: T): void
  clear(): void
}>

/**
 * Context 选择只缓存服务端签发的 opaque contextId/version。版本变化时必须清理宿主查询缓存，
 * 不能把本地选择当成授权，也不接受客户端自报 tenant/project/network 等范围。
 */
export function createContextSelectionStore<T extends ServiceContext>(
  onContextBoundaryChanged: () => void,
): ContextSelectionStore<T> {
  let selected: T | null = null

  return {
    current: () => selected,
    select(next) {
      if (!next.contextId || !next.contextVersion) throw new Error('服务上下文缺少稳定标识或版本')
      if (
        selected &&
        (selected.contextId !== next.contextId || selected.contextVersion !== next.contextVersion)
      ) {
        onContextBoundaryChanged()
      }
      selected = Object.freeze({ ...next }) as T
    },
    clear() {
      if (selected) onContextBoundaryChanged()
      selected = null
    },
  }
}
